tests/core/data/test_tftensor.py

```diff
@@ -0,0 +1,171 @@
+import numpy as np
+from numpy.testing import assert_almost_equal
+import tensorflow as tf
+import pytest
+
+from qutip.core import data
+from qutip.core.data import dense
+from qutip_tensorflow.core.data import TfTensor
+
+from . import conftest
+
+
+# Set up some fixtures for automatic parametrisation.
+
+
+@pytest.fixture(
+    params=[
+        pytest.param((1, 5), id="ket"),
+        pytest.param((5, 1), id="bra"),
+        pytest.param((5, 5), id="square"),
+        pytest.param((2, 4), id="wide"),
+        pytest.param((4, 2), id="tall"),
+    ]
+)
+def shape(request):
+    return request.param
+
+
+@pytest.fixture(params=[True, False], ids=["Fortran", "C"])
+def fortran(request):
+    return request.param
+
+
+def _valid_numpy():
+    # Arbitrary valid numpy array.
+    return conftest.random_numpy_dense((5, 5), False)
+
+
+def _valid_tensor():
+    # Arbitrary valid tensor array.
+    return conftest.random_tensor_dense((5, 5))
+
+
+def _valid_list():
+    # Arbitrary valid tensor array.
+    return conftest.random_numpy_dense((5, 5), False).tolist()
+
+
+@pytest.fixture(scope="function")
+def numpy_dense(shape, fortran):
+    return conftest.random_numpy_dense(shape, fortran)
+
+
+@pytest.fixture(scope="function")
+def tensor_dense(shape):
+    return conftest.random_tensor_dense(shape)
+
+
+@pytest.fixture(scope="function")
+def list_dense(shape):
+    return conftest.random_numpy_dense(shape, fortran).tolist()
+
+
+@pytest.fixture(scope="function")
+def data_tensor_dense(tensor_dense):
+    return TfTensor(tensor_dense)
+
+
+class TestClassMethods:
+    def test_init_from_list(self, list_dense, shape):
+        test = TfTensor(list_dense, shape)
+        assert test.shape == shape
+        assert np.all(test.to_array() == np.array(list_dense))
+
+    def test_init_from_ndarray(self, numpy_dense):
+        test = TfTensor(numpy_dense)
+        assert test.shape == numpy_dense.shape
+        assert np.all(test.to_array() == numpy_dense)
+
+    def test_init_from_tensor(self, tensor_dense):
+        test = TfTensor(tensor_dense)
+        assert test.shape == tuple(tensor_dense.shape.as_list())
+        assert np.all(test.to_array() == tensor_dense)
+
+        # by default we do not return a copy
+        assert test._tf is tensor_dense
```

Comment:
If so, you probably also want to check that using the `copy` keyword _does_ return a copy.